### PR TITLE
Adding checks ensuring modules are not cross-referenced from other module categories

### PR DESCRIPTION
**Pull request by hkaiser**
This makes sure that modules are not linked into more than one shared module library.

### CODE REVIEW

File: cmake/HPX_AddModule.cmake [hunk 1]
```diff
@@ -267,10 +267,33 @@ function(add_hpx_module libname modulename)
...
-  target_link_libraries(
-    hpx_${modulename} PUBLIC ${${modulename}_MODULE_DEPENDENCIES}
-  )
-  target_link_libraries(hpx_${modulename} PUBLIC ${${modulename}_DEPENDENCIES})
+  if(${modulename}_MODULE_DEPENDENCIES)
+    # verify that all dependencies are from the same module category
+    foreach(dep ${${modulename}_MODULE_DEPENDENCIES})
+      # consider only module dependencies, not other targets
+      string(FIND ${dep} "hpx_" find_index)
+      string(FIND ${dep} "hpx_dependencies_" find_dep_index)
```
Comment (msimberg) [inline cmake/HPX_AddModule.cmake:275]: This is a coincidence, but `hpx_dependencies_...` should in fact also not be in `MODULE_DEPENDENCIES`, would you mind changing the check to include that?

File: cmake/HPX_AddModule.cmake [hunk 2]
```diff
@@ -267,10 +267,33 @@ function(add_hpx_module libname modulename)
     # cmake-format: on
   endif()
 
-  target_link_libraries(
-    hpx_${modulename} PUBLIC ${${modulename}_MODULE_DEPENDENCIES}
-  )
-  target_link_libraries(hpx_${modulename} PUBLIC ${${modulename}_DEPENDENCIES})
+  if(${modulename}_MODULE_DEPENDENCIES)
+    # verify that all dependencies are from the same module category
+    foreach(dep ${${modulename}_MODULE_DEPENDENCIES})
```
Comment (msimberg) [inline cmake/HPX_AddModule.cmake:272]: Should we make it optional, and only enable the check on CI? I'd guess it's not terribly slow, but future similar checks may add up...

File: cmake/HPX_AddModule.cmake [hunk 3]
```diff
@@ -267,10 +267,33 @@ function(add_hpx_module libname modulename)
     # cmake-format: on
   endif()
 
-  target_link_libraries(
-    hpx_${modulename} PUBLIC ${${modulename}_MODULE_DEPENDENCIES}
-  )
-  target_link_libraries(hpx_${modulename} PUBLIC ${${modulename}_DEPENDENCIES})
+  if(${modulename}_MODULE_DEPENDENCIES)
```
Comment (msimberg) [inline cmake/HPX_AddModule.cmake:270]: Was this necessary to get things working?

**Review by msimberg — COMMENTED** [review]
Thanks @hkaiser for doing this!

File: cmake/HPX_AddModule.cmake (hunk 3 above)
Comment (hkaiser) [reply cmake/HPX_AddModule.cmake:270]: No, just for consistency purposes. I'm not sure if `target_link_libraries` can be called without arguments.

File: cmake/HPX_AddModule.cmake (hunk 1 above)
Comment (hkaiser) [reply cmake/HPX_AddModule.cmake:275]: I can do that. I will also change the then failing incantations.

File: cmake/HPX_AddModule.cmake (hunk 2 above)
Comment (hkaiser) [reply cmake/HPX_AddModule.cmake:272]: How do I do that? Do we have a flag already in place?

File: cmake/HPX_AddModule.cmake (hunk 3 above)
Comment (msimberg) [reply cmake/HPX_AddModule.cmake:270]: It can indeed be empty, no need to check if it's empty.

File: cmake/HPX_AddModule.cmake (hunk 2 above)
Comment (msimberg) [reply cmake/HPX_AddModule.cmake:272]: No flag yet. We've done these kind of checks separately using shell scripts so far. What I had in mind was just adding an `ADVANCED` flag for this purpose.

File: cmake/HPX_AddModule.cmake (hunk 1 above)
Comment (msimberg) [reply cmake/HPX_AddModule.cmake]: Thank you!

**Comment by hkaiser** [comment]
@msimberg I think I have addressed your comments.

**Review by msimberg — APPROVED** [review]
Looks good now, thank you!

**Pull request merged**